lib/chef/provider/package/dnf.rb

```diff
@@ -0,0 +1,111 @@
+# Copyright:: Copyright 2016, Chef Software, Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef/provider/package"
+require "chef/resource/dnf_package"
+require "chef/mixin/which"
+
+class Chef
+  class Provider
+    class Package
+      class Dnf < Chef::Provider::Package
+        extend Chef::Mixin::Which
+
+        use_multipackage_api
+
+        provides :package, platform_family: %w{rhel fedora} do
+          which("dnf")
+        end
+
+        provides :dnf_package, os: "linux"
+
+        def load_current_resource
+          @current_resource = Chef::Resource::DnfPackage.new(new_resource.name)
+          current_resource.package_name(new_resource.package_name)
+
+          current_resource.version(get_current_versions)
+
+          current_resource
+        end
+
+        def candidate_version
+          @candidate_version ||= begin
+            package_name_array.map do |package_name|
+              provides = dnf("provides", package_name)
```

Comment:
I'm pretty sure we're going to want something like YumCache here (in fact, we can just use YumCache, the metadata for DNF is identical... they still use the same `createrepo`